master/internal/api_runs.go

```diff
@@ -43,3 +56,178 @@ func (a *apiServer) RunPrepareForReporting(
 		StorageId: storageID,
 	}, nil
 }
+
+func (a *apiServer) SearchRuns(
+	ctx context.Context, req *apiv1.SearchRunsRequest,
+) (*apiv1.SearchRunsResponse, error) {
+	resp := &apiv1.SearchRunsResponse{Runs: []*runv1.FlatRun{}}
+	query := db.Bun().NewSelect().
+		Model(&resp.Runs).
+		ModelTableExpr("runs AS r").
+		Apply(getRunsColumns)
+
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get the user: %s", err)
+	}
+	var proj *projectv1.Project
+	if req.ProjectId != nil {
+		proj, err = a.GetProjectByID(ctx, *req.ProjectId, *curUser)
+		if err != nil {
+			return nil, err
+		}
+
+		query = query.Where("e.project_id = ?", req.ProjectId)
+	}
+	if query, err = experiment.AuthZProvider.Get().
+		FilterExperimentsQuery(ctx, *curUser, proj, query,
+			[]rbacv1.PermissionType{rbacv1.PermissionType_PERMISSION_TYPE_VIEW_EXPERIMENT_METADATA},
+		); err != nil {
+		return nil, err
+	}
+
+	if req.Filter != nil {
+		var efr experimentFilterRoot
+		err := json.Unmarshal([]byte(*req.Filter), &efr)
+		if err != nil {
+			return nil, err
+		}
+		query = query.WhereGroup(" AND ", func(q *bun.SelectQuery) *bun.SelectQuery {
+			_, err = efr.toSQL(q)
+			return q
+		}).WhereGroup(" AND ", func(q *bun.SelectQuery) *bun.SelectQuery {
+			if !efr.ShowArchived {
+				return q.Where(`e.archived = false`)
+			}
+			return q
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if req.Sort != nil {
+		err = sortRuns(req.Sort, query)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		query.OrderExpr("id ASC")
+	}
+
+	pagination, err := runPagedBunExperimentsQuery(ctx, query, int(req.Offset), int(req.Limit))
+	resp.Pagination = pagination
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func getRunsColumns(q *bun.SelectQuery) *bun.SelectQuery {
+	return q.
+		Column("r.id").
+		ColumnExpr("proto_time(r.start_time) AS start_time").
+		ColumnExpr("proto_time(r.end_time) AS end_time").
+		ColumnExpr(bunutils.ProtoStateDBCaseString(trialv1.State_value, "r.state", "state",
+			"STATE_")).
+		Column("r.tags").
+		Column("r.checkpoint_size").
+		Column("r.checkpoint_count").
+		Column("r.external_run_id").
+		ColumnExpr(
+			"(SELECT COUNT(*) FROM runs r WHERE e.id = r.experiment_id) AS num_runs").
```

Comment:
this is a weird column to me. Is there a product ask around it?

Reply:
This was initially done so that frontend could know if the run is part of a multi-trial experiment. Instead I have renamed and turned the val into a bool, where it is true if > 1. Makes the field easier to understand and moves the logic off the frontend